### PR TITLE
Multi-Ledger Channel Example

### PR DESCRIPTION
**Pull request by cryptphil**
This PR adds an example of a multi-ledger example for transferring ERC20 tokens between two chains.
Is based on the code of the payment-channel example.

### CODE REVIEW

File: multiledger-channel/main.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+	// The balances that each party puts in the channel from the specific chain.
+	chainABalances := [2]float64{15, 0} // Alice puts 5 PRN from chain A in the channel.
+	chainBBalances := [2]float64{0, 42} // Bob puts 13.37 PRN from chain B in the channel.
+
+	chAlice := alice.OpenChannel(bob.WireAddress(), chainABalances, chainBBalances)
+	chBob := bob.AcceptedChannel()
+
+	log.Println("Sending payments...")
+	chAlice.SendPayment(15, client.ChainAIdx)
+	chBob.SendPayment(42, client.ChainBIdx)
```
Comment (matthiasgeihs) [inline multiledger-channel/main.go:86]: Payments are not a good example of a cross-chain application. Please modify to do a swap.

File: multiledger-channel/client/channel.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+import (
+	"context"
+	"math/big"
+
+	"perun.network/go-perun/channel"
+	"perun.network/go-perun/client"
+)
+
+// PaymentChannel is a wrapper for a Perun channel for the payment use case.
+type PaymentChannel struct {
```
Comment (matthiasgeihs) [inline multiledger-channel/client/channel.go:26]: Payments is not a good example. Better showcase Swaps.

File: multiledger-channel/client/channel.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+// newPaymentChannel creates a new payment channel.
+func newPaymentChannel(ch *client.Channel, currencies [2]channel.Asset) *PaymentChannel {
+	return &PaymentChannel{
+		ch:     ch,
+		assets: currencies,
+	}
+}
+
+// SendPayment sends a payment to the channel peer.
+func (c PaymentChannel) SendPayment(amount float64, chainIndex int) {
```
Comment (matthiasgeihs) [inline multiledger-channel/client/channel.go:40]: Swap instead of payment.

File: multiledger-channel/client/handle.go [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+// HandleUpdate is the callback for incoming channel updates.
+func (c *PaymentClient) HandleUpdate(cur *channel.State, next client.ChannelUpdate, r *client.UpdateResponder) {
+	// We accept every update that increases our balance.
+	err := func() error {
+		err := channel.AssertAssetsEqual(cur.Assets, next.State.Assets)
+		if err != nil {
+			return fmt.Errorf("Invalid assets: %v", err)
+		}
+
+		// Check that our balance is only increasing for all currencies in the channel.
```
Comment (matthiasgeihs) [inline multiledger-channel/client/handle.go:81]: Need to update this for swaps.

File: multiledger-channel/go.mod [hunk 5]
```diff
@@ -0,0 +1,34 @@
+module perun.network/perun-examples/multiledger-channel
+
+go 1.17
+
+require (
+	github.com/ethereum/go-ethereum v1.10.12
+	github.com/perun-network/perun-eth-backend v0.0.0-20220803062824-a30444deaeb8
```
Comment (matthiasgeihs) [inline multiledger-channel/go.mod:7]: Why not use tagged version?

File: multiledger-channel/go.mod [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+	github.com/pkg/errors v0.9.1
+	perun.network/go-perun v0.10.3
```
Comment (matthiasgeihs) [inline multiledger-channel/go.mod:9]: Why not used most recent tag?

**Review by matthiasgeihs — CHANGES_REQUESTED** [review]

File: multiledger-channel/client/client.go [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+		currencies:  assets,
+		channels:    make(chan *PaymentChannel, 1),
+	}
+	go perunClient.Handle(c, c)
+
+	return c, nil
+}
+
+// OpenChannel opens a new channel with the specified peer and funding.
+func (c *PaymentClient) OpenChannel(peer wire.Address, chainABalances, chainBBalances [2]float64) *PaymentChannel {
```
[the remark this answers is not in the document]
Comment (cryptphil) [reply multiledger-channel/client/client.go:136]: I wanted to make it more clear for the tutorial which balance array is for which asset (corresponding to a specific chain). But I'll change it to using `channel.Balances`.

File: multiledger-channel/main.go (hunk 1 above)
Comment (cryptphil) [reply multiledger-channel/main.go]: Modified to a swap.

File: multiledger-channel/client/channel.go (hunk 3 above)
Comment (cryptphil) [reply multiledger-channel/client/channel.go]: Modified to a swap.

File: multiledger-channel/client/handle.go (hunk 4 above)
Comment (cryptphil) [reply multiledger-channel/client/handle.go]: Modified to a swap.

File: multiledger-channel/go.mod (hunk 5 above)
Comment (cryptphil) [reply multiledger-channel/go.mod]: Fixed.

File: multiledger-channel/go.mod (hunk 6 above)
Comment (cryptphil) [reply multiledger-channel/go.mod]: Fixed.

**Review by matthiasgeihs — COMMENTED** [review]
The client struct is still called `PaymentClient` and the channel struct is still called `PaymentChannel`. I thought we agreed to change to `Swap...`?

**Comment by cryptphil** [comment]
> The client struct is still called `PaymentClient` and the channel struct is still called `PaymentChannel`. I thought we agreed to change to `Swap...`?

No, that wasn't clear to me, but it is reasonable; I refactored it.

**Review by matthiasgeihs — APPROVED** [review]

**Pull request merged**